{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6cbb51ec-2467-4d91-8e3a-0db767acf15e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pip\n",
    "\n",
    "def import_or_install(package):\n",
    "    try:\n",
    "        __import__(package)\n",
    "    except ImportError:\n",
    "        pip.main(['install', package]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "95ef8f36-0abf-478c-ace0-d4e43b292c24",
   "metadata": {},
   "outputs": [],
   "source": [
    "packages=['pandas', 'sqlalchemy', 'numpy', 'psycopg2-binary']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2028731b-b26c-4559-b50a-37a6f892f5fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.\n",
      "Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.\n",
      "To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Requirement already satisfied: psycopg2-binary in /opt/homebrew/lib/python3.11/site-packages (2.9.9)\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Requirement already satisfied: psycopg2-binary in /opt/homebrew/lib/python3.11/site-packages (2.9.9)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for package in packages:\n",
    "    import_or_install(package)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "adbbfda5-a716-4abd-b12b-2220bf2c9572",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sqlalchemy import create_engine"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6b59940-0cfb-49b9-afea-563fe77b9c34",
   "metadata": {},
   "source": [
    "Amino Acid Composition\n",
    "This function calculates the percentage composition of each amino acid in a protein sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a23d94a-a657-402c-acba-7f4fb61f796b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def amino_acid_composition(sequence):\n",
    "    from collections import Counter\n",
    "    total_len = len(sequence)\n",
    "    aa_count = Counter(sequence)\n",
    "    composition = {aa: (count / total_len) * 100 for aa, count in aa_count.items()}\n",
    "    return composition"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2f56278-fc95-4fc9-a662-040fab6653d0",
   "metadata": {},
   "source": [
    "Hydrophobicity Scores\n",
    "For this example, we'll use the Kyte & Doolittle scale of hydrophobicity. This function will calculate the average hydrophobicity of the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df60d9bc-208f-4994-ba16-4bbe1e6e7c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_hydrophobicity(sequence):\n",
    "    hydrophobicity_scale = {\n",
    "        'A': 1.8, 'R': -4.5, 'N': -3.5, 'D': -3.5, 'C': 2.5,\n",
    "        'Q': -3.5, 'E': -3.5, 'G': -0.4, 'H': -3.2, 'I': 4.5,\n",
    "        'L': 3.8, 'K': -3.9, 'M': 1.9, 'F': 2.8, 'P': -1.6,\n",
    "        'S': -0.8, 'T': -0.7, 'W': -0.9, 'Y': -1.3, 'V': 4.2\n",
    "    }\n",
    "    scores = [hydrophobicity_scale[aa] for aa in sequence if aa in hydrophobicity_scale]\n",
    "    return sum(scores) / len(scores) if scores else 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87dab0d1-3527-401a-a456-293fb75bb27e",
   "metadata": {},
   "source": [
    "Secondary Structure Prediction Mock-up\n",
    "This is a very simplistic mock model that assigns secondary structures randomly based on typical probabilities (30% helix, 20% sheet, 50% coil)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "94ac70f7-53a4-4e13-96dc-811c8e220790",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def predict_secondary_structure(sequence):\n",
    "    structures = ['Helix', 'Sheet', 'Coil']\n",
    "    probabilities = [0.30, 0.20, 0.50]  # Helix, Sheet, Coil probabilities\n",
    "    return np.random.choice(structures, size=len(sequence), p=probabilities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "69232b36-c6ac-4345-8656-19eeef43579d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the TSV file\n",
    "data = pd.read_csv('../../data/uniparc_AND_database_facet_500_AND_da_2024_05_28.tsv', delimiter='\\t')\n",
    "\n",
    "# Apply the functions to generate additional features\n",
    "data['Amino_Acid_Composition'] = data['Sequence'].apply(amino_acid_composition)\n",
    "data['Avg_Hydrophobicity'] = data['Sequence'].apply(calculate_hydrophobicity)\n",
    "data['Secondary_Structure'] = data['Sequence'].apply(predict_secondary_structure)\n",
    "\n",
    "# Convert complex types from the composition to a string format if needed\n",
    "data['Amino_Acid_Composition'] = data['Amino_Acid_Composition'].apply(str)\n",
    "data['Secondary_Structure'] = data['Secondary_Structure'].apply(lambda x: ','.join(x))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b1a1fb95-1771-4731-81c2-25f9e8b5809c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Organisms</th>\n",
       "      <th>Length</th>\n",
       "      <th>First seen</th>\n",
       "      <th>Last seen</th>\n",
       "      <th>Organism ID</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Pfam</th>\n",
       "      <th>SMART</th>\n",
       "      <th>Amino_Acid_Composition</th>\n",
       "      <th>Avg_Hydrophobicity</th>\n",
       "      <th>Secondary_Structure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UPI00000001D7</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 /...</td>\n",
       "      <td>147</td>\n",
       "      <td>1988-11-01</td>\n",
       "      <td>2024-03-27</td>\n",
       "      <td>559292; 307796; 285006; 574961; 643680; 109563...</td>\n",
       "      <td>Calmodulin;Calmodulin;Calmodulin;Cmd1p;Cmd1p;C...</td>\n",
       "      <td>MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSE...</td>\n",
       "      <td>PF13499</td>\n",
       "      <td>SM00054</td>\n",
       "      <td>{'M': 3.4013605442176873, 'S': 12.244897959183...</td>\n",
       "      <td>-0.294558</td>\n",
       "      <td>Helix,Helix,Sheet,Sheet,Coil,Helix,Sheet,Helix...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UPI0000000336</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 /...</td>\n",
       "      <td>183</td>\n",
       "      <td>1993-07-01</td>\n",
       "      <td>2024-03-27</td>\n",
       "      <td>559292; 307796; 285006; 545124; 574961; 643680...</td>\n",
       "      <td>ARS-binding factor 2, mitochondrial;ARS-bindin...</td>\n",
       "      <td>MNSYSLLTRSFHESSKPLFNLASTLLKASKRTQLRNELIKQGPKRP...</td>\n",
       "      <td>PF00505</td>\n",
       "      <td>SM00398</td>\n",
       "      <td>{'M': 1.092896174863388, 'N': 4.37158469945355...</td>\n",
       "      <td>-0.973224</td>\n",
       "      <td>Coil,Sheet,Coil,Coil,Helix,Sheet,Sheet,Coil,He...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UPI00000003DE</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 /...</td>\n",
       "      <td>203</td>\n",
       "      <td>1995-02-01</td>\n",
       "      <td>2024-03-27</td>\n",
       "      <td>559292; 559292; 307796; 285006; 545124; 574961...</td>\n",
       "      <td>Putative GPI-anchored protein YAR066W;Putative...</td>\n",
       "      <td>MFNRFNKFQAAVALALLSRGALGDSYTNSTSSADLSSITSVSSASA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'M': 0.49261083743842365, 'F': 5.418719211822...</td>\n",
       "      <td>0.081281</td>\n",
       "      <td>Helix,Coil,Coil,Coil,Sheet,Helix,Helix,Coil,Co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UPI0000000521</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 /...</td>\n",
       "      <td>411</td>\n",
       "      <td>1995-02-01</td>\n",
       "      <td>2024-03-27</td>\n",
       "      <td>559292; 4932; 1294345; 559292; 4932; 4932; 559...</td>\n",
       "      <td>Ceramide synthase LAG1;LAG1 isoform 1;Lag1p;sp...</td>\n",
       "      <td>MTSATDKSIDRLVVNAKTRRRNSSVGKIDLGDTVPGFAAMPESAAS...</td>\n",
       "      <td>PF03798</td>\n",
       "      <td>SM00724</td>\n",
       "      <td>{'M': 3.1630170316301705, 'T': 5.1094890510948...</td>\n",
       "      <td>0.123114</td>\n",
       "      <td>Helix,Coil,Coil,Helix,Sheet,Sheet,Sheet,Sheet,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UPI000000056E</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 /...</td>\n",
       "      <td>1367</td>\n",
       "      <td>1995-02-01</td>\n",
       "      <td>2024-03-27</td>\n",
       "      <td>559292; 4932; 4932; 4932; 559292; 4932; 4932; ...</td>\n",
       "      <td>Flocculation protein FLO11;FLO11 isoform 1;FLO...</td>\n",
       "      <td>MQRPFLLAYLVLSLLFNSALGFPTALVPRGSSEGTSCNSIVNGCPN...</td>\n",
       "      <td>PF10182</td>\n",
       "      <td>SM01213</td>\n",
       "      <td>{'M': 0.36576444769568395, 'Q': 1.463057790782...</td>\n",
       "      <td>-0.367740</td>\n",
       "      <td>Helix,Coil,Coil,Helix,Coil,Sheet,Sheet,Helix,H...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Entry                                          Organisms  Length  \\\n",
       "0  UPI00000001D7  Saccharomyces cerevisiae (strain ATCC 204508 /...     147   \n",
       "1  UPI0000000336  Saccharomyces cerevisiae (strain ATCC 204508 /...     183   \n",
       "2  UPI00000003DE  Saccharomyces cerevisiae (strain ATCC 204508 /...     203   \n",
       "3  UPI0000000521  Saccharomyces cerevisiae (strain ATCC 204508 /...     411   \n",
       "4  UPI000000056E  Saccharomyces cerevisiae (strain ATCC 204508 /...    1367   \n",
       "\n",
       "   First seen   Last seen                                        Organism ID  \\\n",
       "0  1988-11-01  2024-03-27  559292; 307796; 285006; 574961; 643680; 109563...   \n",
       "1  1993-07-01  2024-03-27  559292; 307796; 285006; 545124; 574961; 643680...   \n",
       "2  1995-02-01  2024-03-27  559292; 559292; 307796; 285006; 545124; 574961...   \n",
       "3  1995-02-01  2024-03-27  559292; 4932; 1294345; 559292; 4932; 4932; 559...   \n",
       "4  1995-02-01  2024-03-27  559292; 4932; 4932; 4932; 559292; 4932; 4932; ...   \n",
       "\n",
       "                                       Protein names  \\\n",
       "0  Calmodulin;Calmodulin;Calmodulin;Cmd1p;Cmd1p;C...   \n",
       "1  ARS-binding factor 2, mitochondrial;ARS-bindin...   \n",
       "2  Putative GPI-anchored protein YAR066W;Putative...   \n",
       "3  Ceramide synthase LAG1;LAG1 isoform 1;Lag1p;sp...   \n",
       "4  Flocculation protein FLO11;FLO11 isoform 1;FLO...   \n",
       "\n",
       "                                            Sequence     Pfam    SMART  \\\n",
       "0  MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSE...  PF13499  SM00054   \n",
       "1  MNSYSLLTRSFHESSKPLFNLASTLLKASKRTQLRNELIKQGPKRP...  PF00505  SM00398   \n",
       "2  MFNRFNKFQAAVALALLSRGALGDSYTNSTSSADLSSITSVSSASA...      NaN      NaN   \n",
       "3  MTSATDKSIDRLVVNAKTRRRNSSVGKIDLGDTVPGFAAMPESAAS...  PF03798  SM00724   \n",
       "4  MQRPFLLAYLVLSLLFNSALGFPTALVPRGSSEGTSCNSIVNGCPN...  PF10182  SM01213   \n",
       "\n",
       "                              Amino_Acid_Composition  Avg_Hydrophobicity  \\\n",
       "0  {'M': 3.4013605442176873, 'S': 12.244897959183...           -0.294558   \n",
       "1  {'M': 1.092896174863388, 'N': 4.37158469945355...           -0.973224   \n",
       "2  {'M': 0.49261083743842365, 'F': 5.418719211822...            0.081281   \n",
       "3  {'M': 3.1630170316301705, 'T': 5.1094890510948...            0.123114   \n",
       "4  {'M': 0.36576444769568395, 'Q': 1.463057790782...           -0.367740   \n",
       "\n",
       "                                 Secondary_Structure  \n",
       "0  Helix,Helix,Sheet,Sheet,Coil,Helix,Sheet,Helix...  \n",
       "1  Coil,Sheet,Coil,Coil,Helix,Sheet,Sheet,Coil,He...  \n",
       "2  Helix,Coil,Coil,Coil,Sheet,Helix,Helix,Coil,Co...  \n",
       "3  Helix,Coil,Coil,Helix,Sheet,Sheet,Sheet,Sheet,...  \n",
       "4  Helix,Coil,Coil,Helix,Coil,Sheet,Sheet,Helix,H...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6261955d-90bb-40b4-9b47-053c6d0b2a41",
   "metadata": {},
   "source": [
    "Refer [Database Setup](./databaseSetup.md) for steps to set up the postgres database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "79d15110-b4c9-4681-9775-fdf214756517",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Entry                                                         UPI00000001D7\n",
       "Organisms                 Saccharomyces cerevisiae (strain ATCC 204508 /...\n",
       "Length                                                                  147\n",
       "First seen                                                       1988-11-01\n",
       "Last seen                                                        2024-03-27\n",
       "Organism ID               559292; 307796; 285006; 574961; 643680; 109563...\n",
       "Protein names             Calmodulin;Calmodulin;Calmodulin;Cmd1p;Cmd1p;C...\n",
       "Sequence                  MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSE...\n",
       "Pfam                                                                PF13499\n",
       "SMART                                                               SM00054\n",
       "Amino_Acid_Composition    {'M': 3.4013605442176873, 'S': 12.244897959183...\n",
       "Avg_Hydrophobicity                                                -0.294558\n",
       "Secondary_Structure       Helix,Helix,Sheet,Sheet,Coil,Helix,Sheet,Helix...\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6130a344-0fdc-4406-990e-699181d128b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to convert numpy data types to native Python types\n",
    "def convert_types(val):\n",
    "    if isinstance(val, np.integer):\n",
    "        return int(val)\n",
    "    elif isinstance(val, np.floating):\n",
    "        return float(val)\n",
    "    elif isinstance(val, np.ndarray):\n",
    "        return val.tolist()  # Convert arrays to list if needed\n",
    "    else:\n",
    "        return val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7b8a1a77-8c7b-453c-b26e-667279747e63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connect to PostgreSQL\n",
    "engine = create_engine('postgresql://postgres:postgres@localhost:5432/generatebiomedicines')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6a151b11-ccdc-4ec2-a07c-2b5c1853c87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample data \n",
    "sampleDataToInsert = data.iloc[0]\n",
    "\n",
    "# Apply the conversion to all columns in the DataFrame before inserting into SQL\n",
    "sampleDataToInsert = data.map(convert_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9f3e8602-25a8-4106-9765-0d558148be5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to convert dictionary string to JSON\n",
    "def convert_dict_to_json(dict_str):\n",
    "    try:\n",
    "        # Safely evaluate the string to a dictionary and then dump to JSON\n",
    "        return json.dumps(eval(dict_str))\n",
    "    except Exception as e:\n",
    "        print(f\"Error converting to JSON: {e}\")\n",
    "        return None\n",
    "\n",
    "# Function to convert comma-separated list to JSON array\n",
    "def convert_list_to_json(list_str):\n",
    "    try:\n",
    "        # Strip extra quotes and split by comma\n",
    "        list_items = list_str.strip('[]\"').split(',')\n",
    "        list_items = [item.strip().strip('\"') for item in list_items]\n",
    "        return json.dumps(list_items)\n",
    "    except Exception as e:\n",
    "        print(f\"Error converting to JSON: {e}\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d76940f3-5cea-41ee-8d63-58dd8014e0d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "# Insert the first row into the database\n",
    "sampleDataToInsert.columns = [col.lower() for col in data.columns]\n",
    "sampleDataToInsert.columns = [col.lower().replace(' ', '_') for col in data.columns]\n",
    "\n",
    "# Apply the conversion functions to the relevant columns\n",
    "sampleDataToInsert['amino_acid_composition'] = sampleDataToInsert['amino_acid_composition'].apply(convert_dict_to_json)\n",
    "sampleDataToInsert['secondary_structure'] = sampleDataToInsert['secondary_structure'].apply(convert_list_to_json)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "12210a4f-4ba0-4216-89d4-fcbc1ed67221",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\"M\": 3.4013605442176873, \"S\": 12.244897959183673, \"N\": 6.122448979591836, \"L\": 12.244897959183673, \"T\": 2.7210884353741496, \"E\": 12.244897959183673, \"Q\": 4.081632653061225, \"I\": 5.442176870748299, \"A\": 8.16326530612245, \"F\": 5.442176870748299, \"K\": 5.442176870748299, \"D\": 8.16326530612245, \"G\": 5.442176870748299, \"V\": 4.761904761904762, \"R\": 2.0408163265306123, \"P\": 0.6802721088435374, \"H\": 1.3605442176870748}'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDataToInsert.iloc[0]['amino_acid_composition']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e6c52c58-3563-4a0c-adeb-30629caf7e03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\"M\": 3.4013605442176873, \"S\": 12.244897959183673, \"N\": 6.122448979591836, \"L\": 12.244897959183673, \"T\": 2.7210884353741496, \"E\": 12.244897959183673, \"Q\": 4.081632653061225, \"I\": 5.442176870748299, \"A\": 8.16326530612245, \"F\": 5.442176870748299, \"K\": 5.442176870748299, \"D\": 8.16326530612245, \"G\": 5.442176870748299, \"V\": 4.761904761904762, \"R\": 2.0408163265306123, \"P\": 0.6802721088435374, \"H\": 1.3605442176870748}'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "convert_dict_to_json(sampleDataToInsert.iloc[0]['amino_acid_composition'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8df3ba87-e7d8-4d74-827a-1ba5f3c5eb79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "212"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDataToInsert.to_sql('protein_data', con=engine, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ae2aef92-f8fc-490a-a805-748074c77fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetching data for a given entry id from the database\n",
    "def fetchEntryById (entry_id):\n",
    "    query = f\"SELECT * FROM protein_data WHERE entry = '{entry_id}'\"\n",
    "\n",
    "    print(query)\n",
    "\n",
    "    # Execute the query and fetch the row into a DataFrame\n",
    "    data = pd.read_sql(query, engine)\n",
    "    \n",
    "    # Print the fetched row\n",
    "    if not data.empty:\n",
    "        print(\"Fetched Row:\")\n",
    "        print(data)\n",
    "    else:\n",
    "        print(\"No data found for the specified entry.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "59c5c11f-1c13-455e-aba1-53eec1f8a717",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'UPI00000001D7'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleEntryId = sampleDataToInsert.iloc[0][\"entry\"]\n",
    "sampleEntryId"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3d868d0b-4728-4d8d-bdff-109caaddd393",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT * FROM protein_data WHERE entry = 'UPI00000001D7'\n",
      "Fetched Row:\n",
      "           entry                                          organisms  length  \\\n",
      "0  UPI00000001D7  Saccharomyces cerevisiae (strain ATCC 204508 /...     147   \n",
      "\n",
      "   first_seen   last_seen                                        organism_id  \\\n",
      "0  1988-11-01  2024-03-27  559292; 307796; 285006; 574961; 643680; 109563...   \n",
      "\n",
      "                                       protein_names  \\\n",
      "0  Calmodulin;Calmodulin;Calmodulin;Cmd1p;Cmd1p;C...   \n",
      "\n",
      "                                            sequence     pfam    smart  \\\n",
      "0  MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSE...  PF13499  SM00054   \n",
      "\n",
      "                              amino_acid_composition  avg_hydrophobicity  \\\n",
      "0  {\"M\": 3.4013605442176873, \"S\": 12.244897959183...           -0.294558   \n",
      "\n",
      "                                 secondary_structure  \n",
      "0  [\"Helix\", \"Helix\", \"Sheet\", \"Sheet\", \"Coil\", \"...  \n"
     ]
    }
   ],
   "source": [
    "fetchEntryById(sampleEntryId)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
